src/arch/riscv/sbi.c

```diff
@@ -382,6 +389,35 @@ static struct sbiret sbi_hsm_status_handler(void)
     return ret;
 }
 
+static struct sbiret sbi_hsm_suspend_handler(void)
+{
+    struct sbiret ret;
+    uint32_t suspend_type = (uint32_t)vcpu_readreg(cpu()->vcpu, REG_A0);
+    // unsigned long resume_addr = vcpu_readreg(cpu()->vcpu, REG_A1);
+    // unsigned long priv = vcpu_readreg(cpu()->vcpu, REG_A2);
+
+    spin_lock(&cpu()->vcpu->arch.sbi_ctx.lock);
+    if (cpu()->vcpu->arch.sbi_ctx.state != STARTED) {
+        ret.error = SBI_ERR_FAILURE;
+    } else {
+        if (suspend_type & SBI_HSM_SUSP_NON_RET_BIT) {
+            // spin_lock(&cpu()->vcpu->arch.sbi_ctx.lock);
+            // cpu()->vcpu->arch.sbi_ctx.start_addr = resume_addr;
+            // cpu()->vcpu->arch.sbi_ctx.priv = (unsigned int)priv;
+            // spin_unlock(&cpu()->vcpu->arch.sbi_ctx.lock);
+
+            /*
+             * For now we're not supporting non-retentive suspend since we
+             * will need to save and later restore various hart registers and CSRs.
```

Comment:
Can we add a TODO here saying that we only need to implement this when we get a real physical platform with  real non-retentive suspend implemented?

Reply:
> Can we add a TODO here saying that we only need to implement this when we get a real physical platform with real non-retentive suspend implemented?

Done [here](https://github.com/bao-project/bao-hypervisor/compare/5f663e3d4afbf5c9ff81ced7fd05a6a259bba205..170912648fb06cf85cf9853b409920926d3b3987)